sql/distsql: add limits to tablereader

RaduBerinde opened the pull request:
Adding soft and hard limits to TableReader. Improving tests to use a generated
table.

+some test utils in a separate commit.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7041)

<!-- Reviewable:end -->

knz commented:
Reviewed 2 of 2 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/distsql/tablereader.go, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7041#-KJS-8L4mfn_GniMFymC:-KJS-8L5x7Db1Snf1nT1:-303690566) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe122fcecd58f732a8fa69baf08d011d30024a8/sql/distsql/tablereader.go#L92)):*
> ```Go
> 	}
> 
> 	if tr.hardLimit != 0 && tr.hardLimit < tr.softLimit {
> ```

I don't understand why this is an error. Isn't it sufficient to limit at the minimum of the two values?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7041)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> #### sql/distsql: add limits to tablereader
> 
> Adding soft and hard limits to TableReader. Improving tests to use a generated
> 
> table.
> 
> 
> 
> +some test utils in a separate commit.
> 
> 
> 
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7041#-:-KJS-TfRq825Uv6RkBxD:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/distsql/tablereader.go, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7041#-KJS-8L4mfn_GniMFymC:-KJVwualr1ash-xPJn8R:385027027) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fe122fcecd58f732a8fa69baf08d011d30024a8/sql/distsql/tablereader.go#L92)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I don't understand why this is an error. Isn't it sufficient to limit at the minimum of the two values?
</details>
The code that is planning the flow should do that.. From tablereader's perspective, it only makes sense to set both if the soft limit is smaller.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7041#-:-KJVxbpAl5L1wAWm5v5v:-485616692)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: testing code for creating generated tables

Helper code to create tables using "generator" functions for the values.

RaduBerinde committed:
sql/distsql: add limits to tablereader

Adding soft and hard limits to TableReader. Improving tests to use a generated
table.
